### PR TITLE
Implemented a simple i2cslave

### PR DESCRIPTION
**Pull request by seijikun**
I wrote a simple i2cslave who's design was mainly inspired by Arduino's **Wire** library.

Two callbacks `onRecieve` and `onRequest` are fired when data or a request are sent from the master respectively. Those callbacks are called from the `I2CSlave`'s mainloop.
Only one cog is neccessary for a working operation, splitting it up might, however, result in a higher maximal transfer rate. I did only test it with a frequency of `100k`, which worked just fine.

The current coding style of the implementation does not fit with PropWare's, but that's a solvable problem if you can point me in the right direction, I guess.

PropWare is awesome :)

### CODE REVIEW

**Comment by DavidZemon** [comment]
First, thanks for submitting this, and I'm glad you like PropWare! I like this submission a lot, and I'll work with you to get it integrated.

I haven't looked through every line of code yet, but at a quick glance, it looks like this could be an addition to the exist `PropWare::I2C` class rather than a new class. Do you have an concerns with that?

File: Examples/PropWare_I2CSlave/PropWareI2C_Demo.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * <br>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+ * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <PropWare/serial/i2c/i2c.h>
+#include <PropWare/hmi/output/printer.h>
+#include <simpletools.h>
```
Comment (DavidZemon) [inline Examples/PropWare_I2CSlave/PropWareI2C_Demo.cpp]: This inclusion of the Simple library looks unnecessary. Is it? If so, please remove it, as I don't like to require external libraries from within PropWare or its examples.

**Comment by seijikun** [comment]
I don't think it's that easy. Because an I2CMaster has no address, but an I2CSlave does have one. So there would be 2 different constructors (let's say one with and one without the slave address) leading to a possible situation where someone who wants a Slave can use the Constructor that is not taking an address, leading to weird behavior.

Slave and master share quite some functionality, like: Waiting for an ack, sending an ack etc.
However, the implementation of both is completely different, because the master is always giving the clock, whereas the slave is always taking the master's clock. So this could lead to quite some confusion.

But it's your decision in the end :)

**Comment by DavidZemon** [comment]
I was just looking over the `run()` function more closely and coming to the same conclusion. I think this may be why SRLM created an I2CBase class that `libpropeller::I2C` class inherits from. It may be worth looking at `libpropeller::I2CBase`, extracting the same functions out of `PropWare::I2C`, renaming `PropWare::I2C` to `PropWare::I2CMaster`, and then allowing both `PropWare::I2CMaster` and `PropWare::I2CSlave` to inherit from `PropWare::I2CBase`.

**Comment by seijikun** [comment]
The same idea I had. However, when looking for functionality they share - I only found the pin variables for sda and scl.

I originally started by taking the whole I2C class, implementing the slave and throwing everything out I couldn't need. But, well ... I ended up throwing away everything (except those variables).

**Comment by DavidZemon** [comment]
That's too bad. In that case, it may be near perfect. As you mentioned originally, it doesn't fit PropWare's current standards, so we'll want to get that fixed first. I thought I had uploaded my initial work on documenting the standards, but I can't find it. I'll get that [re]uploaded soon and then drop it in here. For the most part, it looks like you did a great job gleaning what you could from other parts of PropWare. What I noticed so far though:
- Methods should be underscore-separated, not camelcase
- No use of dynamic memory allocation, except in very rare circumstances like StringBuilder and the FAT filesystem. Require a pre-allocated array in the constructor instead of a size (see `PropWare::Runnable`'s constructor for [an example](https://github.com/parallaxinc/PropWare/blob/develop/PropWare/concurrent/runnable.h#L108)).

Also, what do you think of inheriting from `PropWare::Runnable`? It seems like a good fit.

**Comment by seijikun** [comment]
What exactly do you have in mind when inheriting of Runnable? Would you make the I2CSlave abstract and force a user to override two methods: `onReceive` and `onRequest` ?
I'm asking because I had virtual methods in a performance critical area (the callbacks are very performance critical), and the virtual call was so terribly slow, that I missed two clock cycles.

This class was created at work. Unforuntately, I have neither a Propeller for private use, nor compiler set up. I therefore can't test and have to do the changes "blindly".
I just committed the change from camelcase to underscore (which I did by hand).
Could you please do the refactoring? I don't feel like I am able to produce a working version without a compiler checking the changes.

**Comment by DavidZemon** [comment]
Sure, I'll pull this into a different branch and then make the changes. I've created issue #102 to track the changes. I'll pull your changes into that branch manually, since (from what I can tell) GitHub does not offer any way to change the target of an existing merge request.

**Comment by seijikun** [comment]
Sounds good!

**Comment by DavidZemon** [comment]
Okay, I made good progress on it today, but ran into a problem during testing. The branch is [here](https://github.com/parallaxinc/PropWare/tree/feature/102) and if you try to run the demo app, you'll notice it fails. It should just print out 0, 1, 2 (it should be echoing the requested register address). So, I'll get out my logic analyzer either tomorrow or Tuesday night, depending on how much time I have tomorrow.

**Comment by seijikun** [comment]
Yeah I had problems getting the I2CMaster to work as well. That is why I tested it with an Arduino board as master instead.
When looking at the logic analyzer, the I2CMaster only sent a start condition directly followed by a stop condition on the bus. I thought that I just did not understand how to use it.

I took the new I2CSlave and am currently testing it with the Arduino. Everything seems to work just right.

Am 7. November 2016 05:10:07 MEZ, schrieb David Zemon notifications@github.com:

> Okay, I made good progress on it today, but ran into a problem during
> testing. The branch is
> [here](https://github.com/parallaxinc/PropWare/tree/feature/102) and if
> you try to run the demo app, you'll notice it fails. It should just
> print out 0, 1, 2 (it should be echoing the requested register
> address). So, I'll get out my logic analyzer either tomorrow or Tuesday
> night, depending on how much time I have tomorrow.
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/parallaxinc/PropWare/pull/101#issuecomment-258743475

## 

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by seijikun** [comment]
Receiving data does not work. This is due to the changes you made to `read_to_end()`.
Moving the variable initialization to the asm code was on purpose, since leaving it in the c++ code did not work for me. (leaving it in asm was the only way for the code to work on every optimization level that I could find)
Most of the things in this function that seem weird first are due to performance reasons. It is very important that this function exits as fast as possible. When resolving the `if/else` condition at the bottom of the while loop, there is a jump and a compare extra, bevore it leaves the function.

I need this library to work at a frequency of 400000. Every single instruction counts at this speed. Could we please make a compromise there?

**Comment by DavidZemon** [comment]
Sure, I'll switch it back. I was planning to test higher speeds later, but was trying to work out that first problem before going back and re-optimizing. I'm happy to see the optimizations are there only because you concrete proof that they need to be.

Unfortunately, I'm in the opposite situation as you. I'm now at work and have no board or compiler to test the changes. I'll make the changes when I get home.

**Comment by DavidZemon** [comment]
Okay, I did some more playing with this. I tried getting my Arduino Leonardo up and running so that I could replicate your tests with it. I couldn't even get the LED to blink lol.

I gave up on that and wrote another test using the I2C master routines that come with the Simple library that yielded different but also failing results. Based on that, I'm thinking there may be something wrong with the I2C slave implementation that is just barely off spec - something that the Arduino allows for but neither PropWare nor Simple do. It's been too long since I looked at the I2C spec, so it's going to take another night of research to brush up and figure it out. Of course, it's equally likely that something is wrong with both Simple and PropWare's I2C master routines, in which case I'm equally interested in getting that fixed. So, in any case, I'm going to avoid merging this into the develop branch until `PropWare::I2CMaster` and `PropWare::I2CSlave` play nicely together.

**Comment by seijikun** [comment]
I see a brutal performance difference between the current version in the `feature/102` branch and the version I originally committed.
In the current branch, it takes the Propeller ~15µs to enter the `read_to_end()` method. Even aggressively inlineing did not seem to improve this.

I set a pin to 1 directly before entering the function, and to 0 as first instruction within the function in both versions. (Optimization-Level: `-Os`) These were the results (@ 200kbit/s):

Original - works:
original
Currently in `feature/102` - does not work due to `read_to_end()` entering too late:
current branch feature/102

Something seems very wrong here. My blind guess would be, that the virtual run method is at fault due to the inheritance from Runnable.
So I took my original version and applied most code-changes you made - except the inheritance from Runnable. Which gave me the performance of the original version: [code](http://pastebin.com/Dhyrx51Y)

But maybe I missed a change that is actually causing this.

**Comment by DavidZemon** [comment]
That's crazy. This certainly deserves some further investigation into why virtual is slowing it down so much. I can understand why virtual would slow down the invocation of a method, but once inside that method I'm stunned to find out that there is a continued performance hit. Hmm......

**Comment by seijikun** [comment]
There seem to be multiple things that can cause this phenomenon. The virtual run method is not at fault.
I started redoing your changes step by step while regularly recompiling and looking at the results in the process. The result of this is now a fast `I2CSlave` inheriting from `Runnable` with nearly all PropWare-ify changes you did.
Unfortunately, I don't have a git client at hand at the moment, so I can't update my forked branch and create a PR until today evening. So I will just paste the code [here](http://pastebin.com/j5qHyeEL) for now.

One of the reasons was, that you added spaces and `" \n\t"` after the jump labels within the asm code.

**Comment by DavidZemon** [comment]
Oh no... I'm really sorry. I went through the same ordeal last night (reverting to your version and then slowly migrating back to mine) and ended up finding that same bug. I but didn't bother commenting here because I thought you'd see the commit log.

**Comment by seijikun** [comment]
I saw the commit, but the version currently in the branch is still regressed for me. I saw only one label (the last label in `read_to_end()` fixed. And I didn't take it serious, that something like this would cause a regression :P

**Comment by DavidZemon** [comment]
@seijikun, did you end up using this object? I tried using it today and found what appears to be a serious bug in the invocation of both the receive and send callbacks.

**Comment by DavidZemon** [comment]
@seijikun, you can disregard my previous comment. I simply didn't look at the demo program at all and therefore was using the object all wrong.

**Pull request merged**